javaslang/src/main/java/javaslang/Lazy.java

```diff
@@ -135,14 +133,11 @@ private Lazy(Supplier<? extends T> supplier) {
      */
     @Override
     public T get() {
-        // using a local var speeds up the double-check idiom by 25%, see Effective Java, Item 71
-        Supplier<? extends T> tmp = supplier;
-        if (tmp != null) {
+        if ((value == null) && (supplier != null)) {
```

Comment:
No, there is no such thing as "implicit volatile". It's just your `volatile` `Supplier` is guarantees you always read the correct `value`.

From the JMM standpoint, you have to argue if there exists an execution that has this relation: read(supplier):null --PO/HB--> read(value):null? It does not, because read(supplier):null is always in SO (and therefore in SW) after write(supplier, null), and write(value, SOME_VALUE) --PO/HB--> write(supplier, null). If there is indeed an execution that does read(value):null, then it violates HB consistency. 

Some might think the whole case looks like the usual publishing piggybacking, regardless that we store the `null` into `supplier`:

Thread 1:
  store A;
  volatile store B; 

Thread 2:
  volatile read B // observes the value from store B
  read A; // guaranteed to observe store A

Glossary: read(field):val -- reading value "val" from "field"; write(field, val) -- writing value "val" to "field"; SO -- synchronization order; SW -- synchronizes-with order; PO -- program order; HB - happens-before order.

I would refer to https://shipilev.net/blog/2014/safe-public-construction/ for tests that actually fail without `volatile`.